### PR TITLE
Reject popup confirmations on close

### LINKED ISSUE #10302: Reject all pending confirmations when notification window closed
We should reject all pending confirmations when the notification window is closed.

We already consider closing the notification window to be a rejection of the confirmation currently being shown. However if the user has multiple pending confirmations, closing the notification window only rejects the first one. This can be confusing the next time the user triggers a new confirmation, because they end up getting presented with the old one instead. This can also be confusing when the user closes the notification window on the lock screen, which currently leaves any confirmations pending without rejecting anything.

Rejecting all pending confirmations upon closing the notification window would further cement the idea that closing the window is a rejection, and it would help prevent the build-up of stale confirmations that are no longer of interest to the user.

### PR DESCRIPTION
**Pull request by ritave**
Fixes: #10302

### CODE REVIEW

**Comment by ritave** [comment]
I have read the CLA Document and I hereby sign the CLA

File: ui/pages/settings/advanced-tab/advanced-tab.container.js [hunk 1]
```diff
@@ -12,7 +12,6 @@ import {
   turnThreeBoxSyncingOnAndInitialize,
   setUseNonceField,
   setIpfsGateway,
-  setLedgerTransportPreference,
```
Comment (rekmarks) [inline ui/pages/settings/advanced-tab/advanced-tab.container.js:15]: The changes in this file seem unrelated to the PR.

**Comment by rekmarks** [comment]
@Gudahtt if all pending confirmations are rejected on notification popup close, how should we handle metrics? The `ApprovalController` probably needs a controller messenger event for this purpose, given the current implementation.

**Comment by Gudahtt** [comment]
Good point.

Right now our "on reject" metrics are in the UI. We can move them into the `tx-state-manager` to preserve them with this approach. And similarly for the "message manager"-style classes, we can have them dispatch metrics on reject as well. The state manager for the message seems like a better place than the UI to dispatch such a metric anyway.

As for the confirmations managed by the approval controller, I am not sure why we'd want to dispatch metrics directly in that controller. It doesn't know what these approvals are for, it doesn't have the context. The one controller using that right now is the permissions controller, so to take that as an example, we could dispatch the rejection metric event in `rejectPermissionsRequest`. That's where we have all of the context about what is being rejected.

File: ui/pages/confirmation/confirmation.js [hunk 2]
```diff
@@ -29,9 +29,9 @@ import { getTemplateValues, getTemplateAlerts } from './templates';
  * immutable and painless. This state is not stored in redux state because it
  * should persist only for the lifespan of the current session, and will only
  * be used on this page. Dismissing alerts for confirmations should persist
- * while the user pages back and forth between confirmations. However, if the
- * user closes the confirmation window and later reopens the extension they
- * should be displayed the alerts again.
+ * while the user pages back and forth between confirmations. If the user
```
Comment (Gudahtt) [inline ui/pages/confirmation/confirmation.js:32]: The alerts discussed here are not the same as the confirmations we're rejecting. These should remain unaffected by the changes in this PR, so this comment change is not needed.

File: ui/components/app/signature-request/signature-request.component.js [hunk 3]
```diff
@@ -34,10 +33,8 @@ export default class SignatureRequest extends PureComponent {
     }
   }
 
-  _beforeUnload = (event) => {
+  _beforeUnload = (_) => {
```
Comment (Gudahtt) [inline ui/components/app/signature-request/signature-request.component.js:36]: Nit: This argument can be omitted entirely since we're not using it

```suggestion
  _beforeUnload = () => {
```

File: ui/components/app/signature-request/signature-request.component.js [hunk 4]
```diff
@@ -17,7 +17,6 @@ export default class SignatureRequest extends PureComponent {
       name: PropTypes.string,
     }).isRequired,
     isLedgerWallet: PropTypes.bool,
-    clearConfirmTransaction: PropTypes.func.isRequired,
```
Comment (Gudahtt) [inline ui/components/app/signature-request/signature-request.component.js:20]: This prop should be removed in `signature-request.container.js` as well, since it's no longer needed. The entire `mapDispatchToProps` function can be omitted, since this is the only function it returns.

File: ui/components/app/signature-request-original/signature-request-original.component.js [hunk 5]
```diff
@@ -55,8 +55,7 @@ export default class SignatureRequestOriginal extends Component {
     this._removeBeforeUnload();
   };
 
-  _beforeUnload = (event) => {
-    const { clearConfirmTransaction, cancel } = this.props;
+  _beforeUnload = (_) => {
```
Comment (Gudahtt) [inline ui/components/app/signature-request-original/signature-request-original.component.js:58]: Nit: This argument can be omitted
```suggestion
  _beforeUnload = () => {
```

File: ui/components/app/dropdowns/network-dropdown.js [hunk 6]
```diff
@@ -68,7 +68,6 @@ function mapDispatchToProps(dispatch) {
         }),
       );
     },
-    hideTestNetMessage: () => dispatch(actions.hideTestNetMessage()),
```
Comment (Gudahtt) [inline ui/components/app/dropdowns/network-dropdown.js:71]: Nit: This also seems unrelated to the other changes in this PR

File: app/scripts/lib/notification-manager.js [hunk 7]
```diff
@@ -12,7 +18,9 @@ export default class NotificationManager {
    */
 
   constructor() {
+    super();
     this.platform = new ExtensionPlatform();
+    extension.windows.onRemoved.addListener(this._onWindowClosed.bind(this));
```
Comment (Gudahtt) [inline app/scripts/lib/notification-manager.js:23]: Nit: We've been using the `ExtensionPlatform` module as a way to encapsulate our extension API interactions, like this one. In keeping with that, this might be better off as an `ExtensionPlatform` method. Something like `addOnRemovedListener`.

Not a blocker, as we aren't even consistently doing that yet today (we use `extensionizer` directly in a few places). But since `this.platform` is already here it might be easy enough.

File: ui/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js [hunk 8]
```diff
@@ -46,21 +46,8 @@ export default class ConfirmEncryptionPublicKey extends Component {
...
 
   _beforeUnload = async (event) => {
-    const {
-      clearConfirmTransaction,
-      cancelEncryptionPublicKey,
-      txData,
-    } = this.props;
-    const { metricsEvent } = this.context;
+    const { cancelEncryptionPublicKey, txData } = this.props;
     await cancelEncryptionPublicKey(txData, event);
```
Comment (Gudahtt) [inline ui/pages/confirm-encryption-public-key/confirm-encryption-public-key.component.js:50]: `cancelEncryptionPublicKey` results in the message being rejected in the background. Since we're doing this automatically on window close now, we don't need to call this anymore.

The same goes for `cancelDecryptMessage` as well, in the `confirm-decrypt-message.component.js` beforeunload handler. Both of these beforeunload handlers are no longer needed.

File: app/scripts/lib/message-manager.js [hunk 9]
```diff
@@ -218,6 +219,13 @@ export default class MessageManager extends EventEmitter {
    *
    */
   rejectMsg(msgId) {
+    this.metricsEvent({
+      event: 'Reject Sig Request',
```
Comment (Gudahtt) [inline app/scripts/lib/message-manager.js:223]: This property corresponds to the "name" option passed into `metricsEvent` in the front end metric event this replaces.

The value for that event was this:

```suggestion
      event: 'Cancel Sig Request Via Notification Close',
```

We should try not to change the event details so that we don't break any reports built on the existing events.

Comment (Gudahtt) [reply app/scripts/lib/message-manager.js:223]: It looks like some of the other event names have changed as well. We'll want to ensure they all match what they were before.

File: app/scripts/background.js [hunk 10]
```diff
@@ -475,6 +477,45 @@ function setupController(initState, initLangCode) {
...
+      controller.encryptionPublicKeyManager.messages
+        .filter((msg) => msg.status === 'unapproved')
+        .forEach((tx) => controller.decryptMessageManager.rejectMsg(tx.id));
+      controller.typedMessageManager.messages
+        .filter((msg) => msg.status === 'unapproved')
+        .forEach((tx) => controller.decryptMessageManager.rejectMsg(tx.id));
+      Object.keys(
+        controller.approvalController.state.pendingApprovals,
+      ).forEach((approvalId) =>
+        controller.approvalController.reject(approvalId, new Error()),
```
Comment (Gudahtt) [inline app/scripts/background.js:512]: This empty error isn't very useful in describing why the request was rejected. This error gets returned to sites awaiting on a response, so we should ensure it matches the provider error spec and that it's informative.

Should we be using the approval controller directly here? (cc @rekmarks ) Right now we're only using this for permission requests, so we could reject each permission request instead via `permissionController.rejectPermissionRequest`, which would set the appropriate error.

In general I don't imagine we'd want to use the approval controller directly, because it's too generic. The controllers making use of it have more context for how to handle rejections and confirmations.

File: app/scripts/lib/message-manager.js (hunk 9 above)
Comment (ritave) [reply app/scripts/lib/message-manager.js:223]: The name of the metric changed because the behaviour changed, it will be fired in more cases and doesn’t represent the same thing happening anymore

For example, it will be fired on pressing cancel button, not only on window close

File: app/scripts/background.js (hunk 10 above)
Comment (ritave) [reply app/scripts/background.js:512]: I'll change it to permissionController for now

File: app/scripts/lib/message-manager.js [hunk 11]
```diff
@@ -217,7 +218,16 @@ export default class MessageManager extends EventEmitter {
    * @param {number} msgId - The id of the Message to reject.
    *
    */
-  rejectMsg(msgId) {
+  rejectMsg(msgId, reason = undefined) {
+    if (reason) {
+      this.metricsEvent({
```
Comment (Gudahtt) [inline app/scripts/lib/message-manager.js:223]: We will need to replicate this pattern in the other signature message managers as well (`personalMessageManager` and `typedMessageManager`). All three of these message managers are used by the signature request confirmations.

File: ui/components/app/signature-request/signature-request.component.js [hunk 12]
```diff
@@ -28,33 +25,6 @@ export default class SignatureRequest extends PureComponent {
...
-      txData: { type },
-    } = this.props;
-    const { metricsEvent } = this.context;
-    metricsEvent({
-      eventOpts: {
-        category: 'Transactions',
-        action: 'Sign Request',
-        name: 'Cancel Sig Request Via Notification Close',
-      },
-      customVariables: {
```
Comment (Gudahtt) [inline ui/components/app/signature-request/signature-request.component.js:50]: Note this extra data  here. This component (`signature-request.component.js`) is used for `eth_signTypedData_v3` and `eth_signTypedData_v4`). So this data should be added in the `typedMessageManager` specifically for types V3 and V4, and not for type V1.

Though honestly the extra data couldn't hurt :thinking: Maybe we can add it for all types. I highly doubt any report is counting on it _not_ being set. These metric events are likely to be replaced with more sensible ones in the not too distant future anyway.

**Review by Gudahtt — COMMENTED** [review]
This mostly looks good! One more set of changes left, for the remaining message managers.

File: ui/components/app/signature-request/signature-request.component.js (hunk 12 above)
Comment (Gudahtt) [reply ui/components/app/signature-request/signature-request.component.js:50]: It looks like this custom variable is still missing from the `typedMessageManager` metric event. We need it to be present at least for v3 and v4 messages, because this is how we distinguish between them on our metrics dashboard.

Comment (Gudahtt) [reply ui/components/app/signature-request/signature-request.component.js:50]: Oh, it looks like I was wrong. This `type` doesn't represent the version at all. It just says that it's a signed type message :disappointed: . Oh well. Sorry about that mixup. I see that you've added it to the metric events already by now though, and that seems useful regardless.

File: app/scripts/background.js [hunk 13]
```diff
@@ -475,6 +481,73 @@ function setupController(initState, initLangCode) {
...
 
+  notificationManager.on(
+    NOTIFICATION_MANAGER_EVENTS.POPUP_CLOSED,
+    rejectUnapprovedNotifications,
+  );
+
+  function rejectUnapprovedNotifications() {
+    // Metamask can open a popup for a login, we need to check if
+    // we're on a confirmation page
+    if (controller.isUnlocked()) {
```
Comment (Gudahtt) [inline app/scripts/background.js:492]: I think we want to reject all confirmations even if the wallet is locked. This is critical for ensuring subsequent `request_account` requests trigger the popup to open.

Issue #10085 is what prompted me to write up the issue this PR is addressing, and that case is about users getting stuck after closing the popup window once while the wallet was locked. Once they do that, subsequent attempts by the site to request access to accounts were rejected because we're already processing that request. Interpreting the closing of the popup as a rejection ensures that the next attempt will open the popup again.

It might be unexpected for a user if this ends up rejecting something unexpectedly that they haven't even seen yet. Maybe there's something we can do to guard against that mistake. I don't really understand why a user would choose not to unlock the wallet in this case though, especially if they've just done something that triggered a confirmation that is important to them.

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by danjm — APPROVED** [review]
